server/src/main/java/org/apache/cassandra/sidecar/job/JobTracker.java

```diff
@@ -0,0 +1,254 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.sidecar.job;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import com.google.common.collect.ImmutableMap;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Tracks and stores the results of long-running jobs running on the sidecar
+ */
+public class JobTracker extends LinkedHashMap<UUID, Job>
```

Comment:
do we really need to expose all the methods from Map here? I only see usages of `getJobsView` , `get`, `computeIfAbsent` (which is not even guarded by the lock), and `put` which seems wrong to have . I think I'd prefer to have a smaller API, that exposes only the things we need. Maybe we can simplify by doing this:

```
        Map<UUID, Job> internalTracker = Collections.synchronizedMap(new LinkedHashMap<UUID, Job>()
        {
            @Override
            protected boolean removeEldestEntry(Map.Entry<UUID, Job> eldest)
            {
                if (size() <= capacity)
                {
                    return false;
                }

                LOGGER.warn("Job tracker reached max size, so expiring job uuid={}", eldest.getKey());
                return true;
            }
        });
```

Reply:
In addition to overriding the LHM, we also intend to make it thread-safe with read/write locks around the operations, so we will need all the methods.

Reply:
How about _composing_ a concurrent hash map, instead of _inheriting_ a map. Composition provides a better flexibility to expose only the necessary APIs. 
I am also not convinced with removing the old jobs. If bounded maps is required, it is clearer to me to reject the creation of new jobs. Otherwise, it (i.e. removing jobs underneath) loses the purpose of tracking. 

```java
public class JobTracker
{
    private ConcurrentHashMap<UUID, Job> jobsById;
```

(somehow my last comment got swallowed. Writing up the same comment again...)

Reply:
Wrapping the map in `Collections.synchronizedMap` makes it thread safe. and we reduce the exposure of the internals

Reply:
I think @yifan-c and I are suggesting the same.

Reply:
The tracker does not need to be a map. It just uses map internally, and the map should be encapsulated within the tracker, not exposing the map APIs.

Reply:
Discussed offline. Simplified tracker API in latest revision